### ISSUE TITLE
OnBatchSubRequest events don't catch exception

### ISSUE THREAD

**Issue by maxgorovenko** [yoanm/php-jsonrpc-server-sdk#79]
I need to check each subrequest if it is granted.

I use OnBatchSubRequestProcessingEvent to inject my code exactly before each subrequest is executed.

The main problem is when access is not granted I need to interrupt only current subrequest. Not the others before or later it. The only way to do this is to throw exception (I created my own AccessDeniedException extends JsonRpcException).
The problem is that this throw breaks all later subrequests, because calling event is outside of the try-catch for method, and the closest try-catch block is outside of foreach subrequests block. So no subrequests called after this.

I belive it would be better to have the opportunity to throw exceptions in before and after subrequest-events and not to cancel others.

P.S. AbstractOnBatchSubRequestProcessEvent is very poor. It would be much better to have there JsonRpcRequest (I need this to get method name for access checking). I did some thick to move this around but it's not looking good.

**Comment by yoanm** [comment]
Hello,

It's on purpose. `OnBatchSubRequestProcessingEvent` is under `Acknowledge` namespace, meaning it's not expected to tweak the request/response from that kind of event, they are made mainly for logging purpose.

From what I understand, you want to check if a specific method is accessible. Why not doing it directly from the PHP method itself in that case ? Right before doing anything else.
With that way you solve all your problems, you can access the method name (just inject it into the php method as constructor argument or better, use the PHP method classname instead of Json-rpc method name to grant the access), you can throw an exception it will be correctly managed by the framework, and bonus, it will work with a basic json-rpc call as well as for json-rpc batch call.

Basically : 
```php
class DummyMethod implements JsonRpcMethodInterface
{
    /**
     * {@inheritdoc}
     */
    public function apply(array $paramList = null)
    {
        // 1. Grant access based on what you want/need (JSON-RPC method name, PHP method classname, http headers, etc)

        // 2. Handle the request
        ...
        // 3. Then return a result
        return [
            'status' => 'done',
        ];
        // Or
        return null;
        // Or
        return 12345;
    }
}
```

**Comment by maxgorovenko** [comment]
It is the only way to do. But I have 50+ methods and I don't want to write the same code  in each. It is like the symfony/security will be called inside each controller-action by hands instead of request events.

**Comment by yoanm** [comment]
> But I have 50+ methods and I don't want to write the same code in each

Totally agree, you should wrap the code inside a class and just pass that class to each methods.
So more something like : 
```php
class DummyMethod implements JsonRpcMethodInterface
{
    private $accesChecker;

    public function __construct($accessChecker)
    {
        $this->accessChecker = $accessChecker;
    }
    /**
     * {@inheritdoc}
     */
    public function apply(array $paramList = null)
    {
        // 1. Grant access based on what you want/need (JSON-RPC method name, PHP method classname, http headers, etc)
        $this->accessChecker->check(.....) // Will throw proper exception in case of of unexpected access
        // 2. Handle the request
        ...
        // 3. Then return a result
        return [
            'status' => 'done',
        ];
        // Or
        return null;
        // Or
        return 12345;
    }
}
```

> It is like the symfony/security will be called inside each controller-action by hands instead of request events.

Glad that you compare that framework to symfony 🙇 , but it is not as advanced as a symfony component (yet 😀). I see your point but what you are looking for is not implemented for the moment (and I currently have no spare time to think about it).

What I would do to have an "automatic access checker" is to implement an abstract class extending `JsonRpcMethodInterface` and then all your JSON-RPC php methods should extend that abstract class.
Something like : 
```php
abstract class MethodWithAccessChecker implements JsonRpcMethodInterface
{
    private $accesChecker;

    public function __construct($accessChecker)
    {
        $this->accessChecker = $accessChecker;
    }

    /**
     * {@inheritdoc}
     */
    public function apply(array $paramList = null)
    {
        // 1. Grant access based on what you want/need (JSON-RPC method name, PHP method classname, http headers, etc)
        $this->accessChecker->check(.....) // Will throw proper exception in case of of unexpected access
        // 2. Handle the request and return response
        return $this->_apply($paramList)        
    }

    abstract public function _apply(array $paramList = null);
}

class DummyMethod extends MethodWithAccessChecker
{
    public function _apply(array $paramList = null)
    {
        // 1. Handle the request
        ...
        // 2. Then return a result
        return [
            'status' => 'done',
        ];
        // Or
        return null;
        // Or
        return 12345;
    }
}
```

**Comment by maxgorovenko** [comment]
Thank you for reply! Abstract MethodWithAccessChecker looks nice ;)

**Comment by yoanm** [comment]
You're welcome 👍 .

_I love that kind of issue fixed so quickly 😄_
